btx/processing/merge.py

```diff
@@ -0,0 +1,262 @@
+import matplotlib.pyplot as plt
+import numpy as np
+import argparse
+import os
+import requests
+
+class StreamtoMtz:
+    
+    """
+    Wrapper for calling various CrystFEL functions for generating a mtz
+    file from a stream file and reporting statistics. The output files,
+    including hkl, mtz, dat, and figures, inherit nomenclature from the
+    input stream file.
+    """
+    
+    def __init__(self, input_stream, symmetry, taskdir, cell):
+        
+        self.stream = input_stream # file of unmerged reflections, str
+        self.symmetry = symmetry # point group symmetry, str
+        self.taskdir = taskdir # path for storing output, str
+        self.cell = cell # pdb or CrystFEL cell file, str
+        self._set_up()
+        
+    def _set_up(self):
+        """
+        Retrieve number of processors to run partialator and the temporary
+        file to which to write the command to sbatch.
+        """
+        # handle parallel logic
+        from mpi4py import MPI
+        comm = MPI.COMM_WORLD
+        self.rank = comm.Get_rank()
+        self.nproc = comm.Get_size()
+        if "NCORES" in os.environ:
+            self.nproc = os.environ['NCORES']
+        
+        # make path to executable
+        if "TMP_EXE" in os.environ:
+            self.tmp_exe = os.environ['TMP_EXE']
+        else:
+            self.tmp_exe = os.path.join(self.taskdir ,f'merge.sh')
+                
+        # generate directories if they don't already exit
+        self.hkl_dir = os.path.join(self.taskdir, "hkl")
+        self.fig_dir = os.path.join(self.taskdir, "figs")
+        if self.rank == 0:
+            for dirname in [self.hkl_dir, self.fig_dir]:
+                os.makedirs(dirname, exist_ok=True)
+
+        # retrieve paths
+        self.script_path = os.path.abspath(__file__)
+        self.python_path = os.environ['WHICHPYTHON']
+        self.prefix = os.path.basename(self.stream).split("stream")[0][:-1]
+        self.outhkl = os.path.join(self.hkl_dir, f"{self.prefix}.hkl")
+
+    def cmd_partialator(self, iterations=1, model='unity', min_res=None, push_res=None):
+        """
+        Write command to merge reflection data using CrystFEL's partialator.
+        https://www.desy.de/~twhite/crystfel/manual-partialator.html.
+        By default scaling and post-refinement are performed. This generates
+        a .mtz file from a .stream file.
```

Comment:
That was already true with `indexamajig` but the elephant in the room is the dependency on crystfel to 1) be available, 2) be the expected version (10.x here if I'm not mistaken).

Maybe we should make these dependencies explicit in the docstring?

Hopefully we manage to containerize things at some point so that this version of the code could be made to match a given version of an docker image, and updates in crystfel could be tracked in a more robust manner.

Reply:
New issue created for this.